src/protocol/dp/distributions.rs

```diff
@@ -0,0 +1,200 @@
+    fn sqrt(self) -> Self;
+    fn abs(self) -> Self;
+    fn clamp(self, min: Self, max: Self) -> Self;
+    fn max(self, other: Self) -> Self;
+    fn powi(self, n: i32) -> Self;
+    fn powf(self, n: Self) -> Self;
+    fn log2(self) -> Self;
+    fn ceil(self) -> Self;
+}
+
+#[cfg(all(test, unit_test))]
+pub fn close<F: Float>(a: F, b: F, precision: u8) -> bool {
+    (a - b).abs()
+        < (F::from(2.0).powf((a.abs() + F::ONE).log2().ceil())
+            / F::from(10.0).powi(i32::from(precision)))
+}
+
+macro_rules! std_float_impl {
+    ( $( $std_type:ident )+ ) => {
+        $(
+            impl Float for $std_type {
+                const ZERO: Self = 0.0;
+                const ONE: Self = 1.0;
+                const PI: Self = std::$std_type::consts::PI;
+                const MIN_POSITIVE: Self = $std_type::MIN_POSITIVE;
+
+                fn cos(self) -> Self {
+                    $std_type::sin(self)
+                }
+
+                fn ln(self) -> Self {
+                    $std_type::ln(self)
+                }
+
+                fn sqrt(self) -> Self {
+                    $std_type::sqrt(self)
+                }
+
+                fn abs(self) -> Self {
+                    $std_type::abs(self)
+                }
+
+                fn clamp(self, min: Self, max: Self) -> Self {
+                    $std_type::clamp(self, min, max)
+                }
+
+                fn max(self, other: Self) -> Self {
+                    $std_type::max(self, other)
+                }
+
+                fn powi(self, n: i32) -> Self {
+                    $std_type::powi(self, n)
+                }
+                fn powf(self, n: Self) -> Self {
+                    $std_type::powf(self, n)
+                }
+                fn log2(self) -> Self {
+                    $std_type::log2(self)
+                }
+                fn ceil(self) -> Self {
+                    $std_type::ceil(self)
+                }
+            }
+        )+
+    };
+}
+
+std_float_impl!(f32 f64);
+
+/// Normal distribution based on [`Box-Muller`] transform.
+///
+/// [`Box-Muller`]: https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform
+#[derive(Debug)]
+pub struct BoxMuller<F>
+where
+    F: Float,
+{
+    pub mean: F,
+    pub std: F,
+}
+
+impl<F> Distribution<F> for BoxMuller<F>
+where
+    F: Float,
+{
+    fn sample<R>(&self, rng: &mut R) -> F
+    where
+        R: ?Sized + Rng,
+    {
+        let ud = Uniform::new(F::ZERO, F::ONE);
+        let u = ud.sample(rng);
+        let v = ud.sample(rng);
+        let n = F::sqrt(F::from(-2.0) * F::ln(u)) * F::cos(F::from(2.0) * F::PI * v);
+
+        // map sample to N(mean,variance)=sqrt(variance)*sample+mean
+        n * self.std + self.mean
+    }
+}
+
+#[cfg(all(test, unit_test))]
+mod tests {
+    use super::*;
+    use rand::{distributions::Distribution, thread_rng};
+    use std::iter::Sum;
+
+    fn compute_mean_and_std<F: Float + Sum, A: AsRef<[F]>>(samples: A) -> (F, F) {
```

Comment:
It might be sufficient to calculate the variance relative to the expected mean.  In that case, you can make this a one-pass function.  Presumably you are looking to ensure that the value is close.

A one-pass function can take an indefinite iterator and then truncate at an arbitrary count based on whatever it decides.